### PR TITLE
Return errors from pil analyzer.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: ast/src/parsed/mod.rs [hunk 1]
```diff
@@ -1144,6 +1144,7 @@ impl<E> Children<E> for IfExpression<E> {
 
 #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Serialize, Deserialize, JsonSchema)]
 pub enum StatementInsideBlock<E = Expression<NamespacedPolynomialReference>> {
+    // TODO add a source ref here.
```
Comment (chriseth) [inline ast/src/parsed/mod.rs:1147]: Created issue here: https://github.com/powdr-labs/powdr/issues/1757

File: pil-analyzer/src/pil_analyzer.rs [hunk 2]
```diff
@@ -125,7 +126,7 @@ impl PILAnalyzer {
         }
     }
 
-    pub fn process(&mut self, mut files: Vec<PILFile>) {
+    pub fn process(&mut self, mut files: Vec<PILFile>) -> Result<(), Vec<Error>> {
```
Comment (chriseth) [inline pil-analyzer/src/pil_analyzer.rs:129]: There are still a lot of panics in functions called from here. Created issue: https://github.com/powdr-labs/powdr/issues/1757

File: pil-analyzer/src/side_effect_checker.rs [hunk 3]
```diff
@@ -54,29 +55,31 @@ impl<'a> SideEffectChecker<'a> {
...
                 self.context = old_context;
                 result
             }
-            Expression::BlockExpression(_, BlockExpression { statements, .. }) => {
+            Expression::BlockExpression(source, BlockExpression { statements, .. }) => {
                 for s in statements {
                     if let StatementInsideBlock::LetStatement(ls) = s {
                         if ls.value.is_none() && self.context != FunctionKind::Constr {
-                            return Err(format!(
+                            // TODO the source location is not exact enough. there should be one for each statement.
```
Comment (chriseth) [inline pil-analyzer/src/side_effect_checker.rs:72]: ref https://github.com/powdr-labs/powdr/issues/1757

File: pil-analyzer/tests/condenser.rs [hunk 4]
```diff
@@ -1,9 +1,24 @@
...
+use powdr_ast::analyzed::Analyzed;
 use powdr_number::GoldilocksField;
-use powdr_pil_analyzer::analyze_string;
 use test_log::test;
 
 use pretty_assertions::assert_eq;
 
+fn analyze_string(input: &str) -> Analyzed<GoldilocksField> {
+    powdr_pil_analyzer::analyze_string(input)
+        .map_err(|errors| {
```
Comment (chriseth) [inline pil-analyzer/tests/condenser.rs:10]: This code is duplicated in the other test files, but it's very difficult to share code across test binaries, so I thought it's fine.

File: executor/src/witgen/block_processor.rs [hunk 5]
```diff
@@ -150,7 +150,14 @@ mod tests {
...
     ) -> R {
-        let analyzed = analyze_string(src);
+        let analyzed = analyze_string(src)
+            .map_err(|errors| {
+                for e in errors {
+                    e.output_to_stderr();
+                }
+                panic!("Failed to analyze test input.");
+            })
+            .unwrap();
```
Comment (Schaeff) [inline executor/src/witgen/block_processor.rs:160]: This is panicking twice? Could use `expect` instead of `unwrap`?

**Review by Schaeff — APPROVED** [review]

**Pull request merged**